### DISCUSSION TITLE
More api endpoints

### DISCUSSION THREAD

**Discussion by bobo** [teslamate-org/teslamate#456] [Ideas]
Whats your thoughts on adding more api endpoints?

instead of using tasker etc i made an android app with 2 buttons to resume/suspend polling. And there it would be nice to see the current status to know if its polling or not.

Would also be nice to maybe create a widget etc to show current status of the car without having to call the tesla api.

This could ofc be done with another service building state through mqtt, but i thought i should check what your thoughs are before doing anything :-)

**Comment by rfvgyhn** [comment]
I had a similar desire. I also started to work on an android app and showing some basic state was something I wanted. I ended up losing focus and working on other things so I never made it past my mock api though.

![teslamate-mobile](https://user-images.githubusercontent.com/731861/74596749-c26f6300-5010-11ea-86c8-edcfa2b64a4a.png)

**Comment by adriankumpf** [comment]
> Whats your thoughts on adding more api endpoints?

An app would definitely be great! If all you need is more API endpoints, we should add them.

Do you have any specific endpoints in mind? For example, `GET /api/car/1/summary` could provide all values that are already published via MQTT.


```json
{
  "battery_level": 51,
  "charge_energy_added": 39.57,
  "charge_limit_soc": 90,
  "charge_port_door_open": false,
  "charger_actual_current": 0,
  "charger_phases": null,
  "charger_power": 0,
  "charger_voltage": 1,
  "display_name": "$name",
  "est_battery_range_km": 212.63,
  "geofence": null,
  "heading": 101,
  "healthy": true,
  "ideal_battery_range_km": 252.38,
  "inside_temp": 19.0,
  "is_climate_on": true,
  "is_preconditioning": false,
  "is_user_present": true,
  "latitude": 0.000000,
  "locked": true,
  "longitude": 0.000000,
  "odometer": 26089.63,
  "outside_temp": 10.5,
  "plugged_in": false,
  "rated_battery_range_km": 252.38,
  "scheduled_charging_start_time": null,
  "sentry_mode": false,
  "shift_state": "D",
  "since": "2020-02-16T11:58:05.741490Z",
  "speed": 51,
  "state": "driving",
  "time_to_full_charge": 0.0,
  "update_available": false,
  "usable_battery_level": 51,
  "version": "2020.4.1",
  "windows_open": false
}
```

**Comment by bobo** [comment]
Not sure exactly what endpoints i would need. But i can create them myself as i figure it out. Summary for example is probably one of them atleast. 
Thanks for the response.

**Comment by smefa** [comment]
It would be awesome to be able to check current status through an API. Voting for this one!

**Comment by theskyisthelimit** [comment]
It would be very cool to have an API to get all those informations. Imagine the possibilities!

**Comment by houseofdreams** [comment]
> > Whats your thoughts on adding more api endpoints?
> 
> An app would definitely be great! If all you need is more API endpoints, we should add them.
> 
> Do you have any specific endpoints in mind? For example, `GET /api/car/1/summary` could provide all values that are already published via MQTT.
> 
> ```json
> {
>   "battery_level": 51,
>   "charge_energy_added": 39.57,
>   "charge_limit_soc": 90,
>   "charge_port_door_open": false,
>   "charger_actual_current": 0,
>   "charger_phases": null,
>   "charger_power": 0,
>   "charger_voltage": 1,
>   "display_name": "$name",
>   "est_battery_range_km": 212.63,
>   "geofence": null,
>   "heading": 101,
>   "healthy": true,
>   "ideal_battery_range_km": 252.38,
>   "inside_temp": 19.0,
>   "is_climate_on": true,
>   "is_preconditioning": false,
>   "is_user_present": true,
>   "latitude": 0.000000,
>   "locked": true,
>   "longitude": 0.000000,
>   "odometer": 26089.63,
>   "outside_temp": 10.5,
>   "plugged_in": false,
>   "rated_battery_range_km": 252.38,
>   "scheduled_charging_start_time": null,
>   "sentry_mode": false,
>   "shift_state": "D",
>   "since": "2020-02-16T11:58:05.741490Z",
>   "speed": 51,
>   "state": "driving",
>   "time_to_full_charge": 0.0,
>   "update_available": false,
>   "usable_battery_level": 51,
>   "version": "2020.4.1",
>   "windows_open": false
> }
> ```

Hi, I'm currently making some sort of custom dashboard, but I can't seem to find some settings that I want to add...

You gave an example GET /api/car/1/summary, with exactly the sort of data I'm missing in the current postgresql database. (locked status, sentry mode status, ...)

You mentioned that this data is already published via MQTT. How does this work? Where can I access this data?

Thanks!

**Comment by theskyisthelimit** [comment]
> > Whats your thoughts on adding more api endpoints?
> 
> An app would definitely be great! If all you need is more API endpoints, we should add them.
> 
> Do you have any specific endpoints in mind? For example, `GET /api/car/1/summary` could provide all values that are already published via MQTT.
> 
> ```json
> {
>   "battery_level": 51,
>   "charge_energy_added": 39.57,
>   "charge_limit_soc": 90,
>   "charge_port_door_open": false,
>   "charger_actual_current": 0,
>   "charger_phases": null,
>   "charger_power": 0,
>   "charger_voltage": 1,
>   "display_name": "$name",
>   "est_battery_range_km": 212.63,
>   "geofence": null,
>   "heading": 101,
>   "healthy": true,
>   "ideal_battery_range_km": 252.38,
>   "inside_temp": 19.0,
>   "is_climate_on": true,
>   "is_preconditioning": false,
>   "is_user_present": true,
>   "latitude": 0.000000,
>   "locked": true,
>   "longitude": 0.000000,
>   "odometer": 26089.63,
>   "outside_temp": 10.5,
>   "plugged_in": false,
>   "rated_battery_range_km": 252.38,
>   "scheduled_charging_start_time": null,
>   "sentry_mode": false,
>   "shift_state": "D",
>   "since": "2020-02-16T11:58:05.741490Z",
>   "speed": 51,
>   "state": "driving",
>   "time_to_full_charge": 0.0,
>   "update_available": false,
>   "usable_battery_level": 51,
>   "version": "2020.4.1",
>   "windows_open": false
> }
> ```

I am very interested in an API which would give back all the data as a json file. When accessing `/api/car/1/summary` I get an 404 error. So this feature is not yet implemented, correct?

**Comment by tobiasehlert** [comment]
@theskyisthelimit @houseofdreams @smefa @Rfvgyhn @bobo 

I've released TeslaMateApi (RestAPI container) which might be of interest to you to check out.
With that you can retrieve data just as what you are looking for above :)

https://github.com/tobiasehlert/teslamateapi

Regards,
Tobias

**Reply by theskyisthelimit** [reply]
This is amazing! Works perfectly. Thank you!!!!

**Comment by ToniA** [comment]
I think there should be a wake-up API. Of course you could also directly call the Tesla API, and it's not hard to do with cURL. The thing is that Teslamate already knows what the access token is, and also you'd probably want to start Teslamate logging at the same time.

I also think there should be an option to tell for how long you'd like to have the car online. This can be implemented in Teslamate, just by having it polling the Tesla vehicle details API.

**Reply by tobiasehlert** [reply]
There is a PR in the TeslaMateApi service containing possibility to do a wake_up command towards your car. :)

https://github.com/tobiasehlert/teslamateapi/tree/feature-commands

**Reply by ToniA** [reply]
Cool, the wake-up command is exactly what I need.

Any suggestions on how to handle this 'keep online' thing? If I just wake it up, it stays online for some 10-15 minutes. If I'd like it to stay online for a longer period of time, the best I know at the moment is to repeatedly call Teslamate's resume endpoint, like `curl -X PUT http://localhost:4000/api/car/1/logging/resume`
